### ISSUE TITLE
No optimization output of SPO+ and DBB

### ISSUE THREAD

**Issue by zhechen003** [khalil-research/PyEPO#50]
In the example code of the Training part, the SPO+ function and DBB function only have one loss output. From the source code, the SPO+ and DBB also only return loss but no optimization solution. I think we also need the optimization solution.  This is the SPO+ forward function.
```python
    def forward(ctx, pred_cost, true_cost, true_sol, true_obj, module):
        """
        Forward pass for SPO+

        Args:
            pred_cost (torch.tensor): a batch of predicted values of the cost
            true_cost (torch.tensor): a batch of true values of the cost
            true_sol (torch.tensor): a batch of true optimal solutions
            true_obj (torch.tensor): a batch of true optimal objective values
            module (optModule): SPOPlus modeul

        Returns:
            torch.tensor: SPO+ loss
        """
        # get device
        device = pred_cost.device
        # convert tenstor
        cp = pred_cost.detach().to("cpu").numpy()
        c = true_cost.detach().to("cpu").numpy()
        w = true_sol.detach().to("cpu").numpy()
        z = true_obj.detach().to("cpu").numpy()
        # check sol
        #_check_sol(c, w, z)
        # solve
        sol, obj = _solve_or_cache(2 * cp - c, module)
        # calculate loss
        loss = []
        for i in range(len(cp)):
            loss.append(- obj[i] + 2 * np.dot(cp[i], w[i]) - z[i])
        # sense
        if module.optmodel.modelSense == EPO.MINIMIZE:
            loss = np.array(loss)
        if module.optmodel.modelSense == EPO.MAXIMIZE:
            loss = - np.array(loss)
        # convert to tensor
        loss = torch.FloatTensor(loss).to(device)
        sol = torch.FloatTensor(sol).to(device)
        # save solutions
        ctx.save_for_backward(true_sol, sol)
        # add other objects to ctx
        ctx.optmodel = module.optmodel
        return loss
```

**Comment by LucasBoTang** [comment]
Hi @zhechen003 ,

Thanks for raising this point! First, to clarify, the `DBB` function (Differentiable Black-Box Optimizer) does return only the predicted solutions, not the loss. You can confirm this in the [DBB forward function implementation](https://github.com/khalil-research/PyEPO/blob/main/pkg/pyepo/func/blackbox.py#L57):

```python
class blackboxOptFunc(Function):
    """
    A autograd function for differentiable black-box optimizer
    """

    @staticmethod
    def forward(ctx, pred_cost, module):
        """
        Forward pass for DBB

        Args:
            pred_cost (torch.tensor): a batch of predicted values of the cost
            module (optModule): blackboxOpt module

        Returns:
            torch.tensor: predicted solutions
        """
        # get device
        device = pred_cost.device
        # convert tensor
        cp = pred_cost.detach().to("cpu").numpy()
        # solve
        sol, _ = _solve_or_cache(cp, module)
        # convert to tensor
        sol = torch.tensor(sol, dtype=torch.float, device=device)
        # save
        ctx.save_for_backward(pred_cost, sol)
        # add other objects to ctx
        ctx.lambd = module.lambd
        ctx.module = module
        return sol
```

**Comment by LucasBoTang** [comment]
Regarding the `SPO+` loss, it is true that only the loss is returned, and the optimal solution for the predicted costs `cp` is and will be not included in the output. Let me explain the reasoning behind this design:
1. During the computation of the `SPO+` loss and its corresponding subgradient, the optimal solution for `cp` is not required.
2. Computing the optimal solution for `cp` involves solving an optimization problem, which is a computational bottleneck in training. Adding this extra step in the forward function would significantly increase the computational overhead, especially in large-scale applications.

**Comment by LucasBoTang** [comment]
If youneed to compute the solution explicitly, you can easily do so by leveraging the `optmodel` object directly during training. Here’s an example:

```python
# training
num_epochs = 10
for epoch in range(num_epochs):
    for data in dataloader:
        x, c, w, z = data
        # forward pass
        cp = predmodel(x)
        loss = spop(cp, c, w, z)
        # use optmodel to get solution
        sols = []
        for cp_j in cp.to("cpu").detach().numpy():
            optmodel.setObj(cp_j)
            sol, _ = optmodel.solve()
            sols.append(sol)
        # backward pass
        optimizer.zero_grad()
        loss.backward()
        optimizer.step()
```

**Comment by LucasBoTang** [comment]
I hope the explanation and example provided address your question. If there’s anything still unclear or if you have any additional questions, please feel free to let us know.

If there are no further concerns, I will plan to close this issue after one day.

Looking forward to your feedback!

**Comment by zhechen003** [comment]
Thank you for your explanation! I will carefully reread the example code. Thank you again for your efforts!
